llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -1329,6 +1329,38 @@ SDValue DAGCombiner::reassociateReduction(unsigned RedOpc, unsigned Opc,
                        DAG.getNode(Opc, DL, N0.getOperand(0).getValueType(),
                                    N0.getOperand(0), N1.getOperand(0)));
   }
+
+  // Reassociate add(add(vecreduce(a), b), add(vecreduce(c), d)) into
+  // add(vecreduce(add(a, c)), add(b, d)), to combine the reductions into a
+  // single node.
+  if (N0.getOpcode() == Opc && N1.getOpcode() == Opc && N0->hasOneUse() &&
+      N1->hasOneUse()) {
+    SDValue N00 = N0.getOperand(0);
+    SDValue N01 = N0.getOperand(1);
+    if (N00.getOpcode() != RedOpc && N01.getOpcode() == RedOpc)
+      std::swap(N00, N01);
+    if (N00.getOpcode() == RedOpc && N01.getOpcode() != RedOpc &&
+        N00->hasOneUse()) {
```

Comment:
I'm pretty sure you can use sd_match to handle all of this matching (including the operand swapping).

Reply:
Oh that does sound like a better approach. Thanks.